lib/Dialects/VectorExt/Transform/VectorMaskingUtils.cpp

```diff
@@ -0,0 +1,138 @@
+//===- VectorMaskingUtils.cpp - Utilities for vector masking --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Dialects/VectorExt/VectorMaskingUtils.h"
+
+#include "Dialects/VectorExt/VectorExtOps.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallPtrSet.h"
+
+using namespace mlir;
+using namespace mlir::linalg;
+using namespace mlir::vector;
+using namespace mlir::vector_ext;
+
+using ExprList = ArrayRef<ArrayRef<AffineExpr>>;
+
+/// Move the operation range `opRange` before operation `dest`.
+static void moveOperationsBefore(llvm::iterator_range<Block::iterator> opRange,
+                                 Operation *dest) {
+  if (opRange.empty())
+    return;
+  dest->getBlock()->getOperations().splice(
+      Block::iterator(dest), opRange.begin()->getBlock()->getOperations(),
+      opRange.begin(), opRange.end());
+}
+
+/// Iterate over all the definitions produced by the operation range `opRange`
+/// and gather those with users outside the range.
+static void
+getDefsWithExternalUses(llvm::iterator_range<Block::iterator> opRange,
+                        SmallVectorImpl<Value> &defsWithExternalUses,
+                        SmallVectorImpl<Type> &defTypes) {
+  SmallPtrSet<Operation *, 8> rangeSet;
+  for (Operation &op : opRange)
+    rangeSet.insert(&op);
+
+  for (Operation &op : llvm::make_early_inc_range(opRange))
+    for (Value res : op.getResults())
+      for (Operation *user : res.getUsers())
+        if (rangeSet.count(user) == 0) {
+          defsWithExternalUses.push_back(res);
+          defTypes.push_back(res.getType());
+          break;
+        }
+}
+
+/// Introduce a vector.predicate op that encloses the loop body of the tiled
+/// loop. The predicate is computed based on the provided vectorization factor.
+static void createVectorPredicateOp(TiledLoopOp loopOp, int64_t vecFactor) {
+  // Compute the range of operations that will be moved within vector.predicate
+  // and the definitions within the range with external users.
+  Block &bodyBlock = loopOp.getLoopBody().front();
+  auto opsToMove =
+      llvm::make_range(bodyBlock.begin(), Block::iterator(bodyBlock.back()));
+
+  SmallVector<Value, 8> defsWithExternalUses;
+  SmallVector<Type, 8> defTypes;
+  getDefsWithExternalUses(opsToMove, defsWithExternalUses, defTypes);
+
+  // Initialize a builder to insert vector.predicate at the beginning of the
+  // loop body.
+  Location loc = loopOp.getLoc();
+  MLIRContext *context = loopOp.getContext();
+  OpBuilder builder(loopOp);
```

Comment:
Passing the builder as a parameter and using RAII InsertionGuard + setting insertion points is preferred.
This is a prerequisite for proper composition of transformations and reuse in patterns (if applicable).
We might as well put it in this form from the beginning and avoid artificial frictions on composability.

Reply:
Done. At least in LLVM, the builder is intentionally lightweight to encourage its arbitrary creation/destruction as needed and avoid to carry a single builder all over the place. I guess you are referring to being able to pass either a builder or a rewriter, right?

Reply:
yes, it's to avoid creating a builder internally because that would mess up a PatternRewriter user of the API.